internal/server/singleprocess/service_runner.go

```diff
@@ -93,16 +245,64 @@ func (s *service) RunnerConfig(
 		return err
 	}
 
-	// Defer deleting this.
-	// TODO(mitchellh): this is too aggressive and we want to have some grace
-	// period for reconnecting clients. We should clean this up.
+	// Mark the runner as offline if they disconnect from the config stream loop.
 	defer func() {
-		log.Trace("deleting runner")
-		if err := s.state.RunnerDelete(record.Id); err != nil {
-			log.Error("failed to delete runner data. This should not happen.", "err", err)
+		log.Trace("marking runner as offline")
+		if err := s.state.RunnerOffline(record.Id); err != nil {
+			log.Error("failed to mark runner as offline. This should not happen.", "err", err)
 		}
 	}()
 
+	// We'll set preadopt to true if we're in a state where we can implicitly
+	// adopt the runner if we don't have explicit adoption. This is allowed
+	// in two cases: (1) legacy login tokens for pre WP 0.8 and (2) manually
+	// created runner tokens that match this runner.
+	preadopt := false
```

Comment:
By my reading of the logic, it's not possible to get to line 290 without setting `preadopt` so do we need it? We're rejecting all paths that don't set `preadopt` it appears.

Reply:
I thought about this overnight and agree.